frame/support/src/storage/types/counted_nmap.rs

```diff
@@ -0,0 +1,1446 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Storage map type. Implements CountedStorageNMap, CountedStorageIterableNMap,
+//! CountedStoragePrefixedNMap traits and their methods directly.
+
+use crate::{
+	metadata::StorageEntryMetadata,
+	storage::{
+		types::{
+			EncodeLikeTuple, HasKeyPrefix, HasReversibleKeyPrefix, OptionQuery, QueryKindTrait,
+			StorageEntryMetadataBuilder, StorageNMap, StorageValue, TupleToEncodedIter, ValueQuery,
+		},
+		KeyGenerator, PrefixIterator, StorageAppend, StorageDecodeLength,
+	},
+	traits::{Get, GetDefault, StorageInfo, StorageInstance},
+	Never,
+};
+use codec::{Decode, Encode, EncodeLike, FullCodec, MaxEncodedLen, Ref};
+use sp_runtime::traits::Saturating;
+use sp_std::prelude::*;
+
+/// A wrapper around a `StorageNMap` and a `StorageValue<Value=u32>` to keep track of how many items
+/// are in a map, without needing to iterate over all of the values.
+///
+/// This storage item has some additional storage read and write overhead when manipulating values
+/// compared to a regular storage map.
+///
+/// For functions where we only add or remove a value, a single storage read is needed to check if
+/// that value already exists. For mutate functions, two storage reads are used to check if the
+/// value existed before and after the mutation.
+///
+/// Whenever the counter needs to be updated, an additional read and write occurs to update that
+/// counter.
+pub struct CountedStorageNMap<
+	Prefix,
+	Key,
+	Value,
+	QueryKind = OptionQuery,
+	OnEmpty = GetDefault,
+	MaxValues = GetDefault,
+>(core::marker::PhantomData<(Prefix, Key, Value, QueryKind, OnEmpty, MaxValues)>);
+
+/// The requirement for an instance of [`CountedStorageNMap`].
+pub trait CountedStorageNMapInstance: StorageInstance {
+	/// The prefix to use for the counter storage value.
+	type CounterPrefix: StorageInstance;
+}
+
+// Private helper trait to access map from counted storage double map
+trait MapWrapper {
+	type Map;
+}
+
+impl<P: CountedStorageNMapInstance, K, V, Q, O, M> MapWrapper
+	for CountedStorageNMap<P, K, V, Q, O, M>
+{
+	type Map = StorageNMap<P, K, V, Q, O, M>;
+}
+
+type CounterFor<P> =
+	StorageValue<<P as CountedStorageNMapInstance>::CounterPrefix, u32, ValueQuery>;
+
+/// On removal logic for updating counter while draining upon some prefix with
+/// [`crate::storage::PrefixIterator`].
+pub struct OnRemovalCounterUpdate<Prefix>(core::marker::PhantomData<Prefix>);
+
+impl<Prefix: CountedStorageNMapInstance> crate::storage::PrefixIteratorOnRemoval
+	for OnRemovalCounterUpdate<Prefix>
+{
+	fn on_removal(_key: &[u8], _value: &[u8]) {
+		CounterFor::<Prefix>::mutate(|value| value.saturating_dec());
+	}
+}
+
+impl<Prefix, Key, Value, QueryKind, OnEmpty, MaxValues>
+	crate::storage::generator::StorageNMap<Key, Value>
+	for CountedStorageNMap<Prefix, Key, Value, QueryKind, OnEmpty, MaxValues>
```

Comment:
I think this is not correct: see this example fails:
```diff
diff --git a/frame/support/src/storage/types/counted_nmap.rs b/frame/support/src/storage/types/counted_nmap.rs
index 90edc7334c4..e8e92131aed 100644
--- a/frame/support/src/storage/types/counted_nmap.rs
+++ b/frame/support/src/storage/types/counted_nmap.rs
@@ -693,6 +693,36 @@ mod test {
                }
        }
 
+       #[test]
+       fn gui_test_1() {
+               type A = CountedStorageNMap<Prefix, NMapKey<Blake2_128Concat, u16>, u32, OptionQuery>;
+               type AValueQueryWithAnOnEmpty =
+                       CountedStorageNMap<Prefix, NMapKey<Blake2_128Concat, u16>, u32, ValueQuery, ADefault>;
+               type B = CountedStorageNMap<Prefix, NMapKey<Blake2_256, u16>, u32, ValueQuery>;
+               type C = CountedStorageNMap<Prefix, NMapKey<Blake2_128Concat, u16>, u8, ValueQuery>;
+               type WithLen = CountedStorageNMap<Prefix, NMapKey<Blake2_128Concat, u16>, Vec<u32>>;
+
+               TestExternalities::default().execute_with(|| {
+                       let mut k: Vec<u8> = vec![];
+                       k.extend(&twox_128(b"test"));
+                       k.extend(&twox_128(b"Foo"));
+                       k.extend(&3u16.blake2_128_concat());
+                       assert_eq!(A::hashed_key_for((&3,)).to_vec(), k);
+
+                       assert_eq!(A::contains_key((3,)), false);
+                       assert_eq!(A::get((3,)), None);
+                       assert_eq!(AValueQueryWithAnOnEmpty::get((3,)), 98);
+                       assert_eq!(A::count(), 0);
+
+                       <A as crate::storage::StorageNMap<_, _>>::insert((3,), 10);
+                       assert_eq!(A::contains_key((3,)), true);
+                       assert_eq!(A::get((3,)), Some(10));
+                       assert_eq!(AValueQueryWithAnOnEmpty::get((3,)), 10);
+                       assert_eq!(A::count(), 1);
+               })
+       }
+
+
        #[test]
        fn test_1_key() {
                type A = CountedStorageNMap<Prefix, NMapKey<Blake2_128Concat, u16>, u32, OptionQuery>;
```

When you implement the generator it means that the trait `frame_support::storage::StorageNMap` is implemented using your implementation of the generator.  The implementation of `frame_support::storage::StorageNMap` is in `frame/support/src/storage/generator/nmap.rs:177`. This implementation makes no regards to the counter.

Reply:
fixed thanks